logging/logfile_test.go

```diff
@@ -117,6 +118,110 @@ func TestLogFile_FileRotation_Disabled(t *testing.T) {
+			// input filename has a non-.log extension.
+			// For example:
+			// "consul.txt" => "consul-1629417801659820807.txt"
+			inputFilename: "consul.txt",
+			outputFilenamePattern: `consul-\d+.txt`,
+		},
+		{
+			desc: "template with human readable timestamp",
+			// Test whether a human-readable timestamp is injected at the location
+			// specified in the template.
+			// For example:
+			// "con-${timestamp-human}-sul" => "con-2021-08-19_20-13-08-322108785-sul.log"
+			inputFilename: "con-${timestamp-human}-sul",
+			outputFilenamePattern: `con-\d{4}-\d{2}-\d{2}_\d{2}-\d{2}-\d{2}-\d{9}-sul.log`,
+		},
+		{
+			desc: "template with unix timestamp",
+			// Test whether a human-readable timestamp is injected at the location
+			// specified in the template.
+			// For example:
+			// "${timestamp-unix}-consul" => "1629418506846614551-consul.log"
+			inputFilename: "${timestamp-unix}-consul",
+			outputFilenamePattern: `\d+-consul.log`,
+		},
+		{
+			desc: "template with multiple fields",
+			// Test whether multiple fields can be injected. This isn't expected to be
+			// a real use case, but we want to test this extreme case.
+			// For example:
+			// "${timestamp-unix}-con${timestamp-unix}sul-${timestamp-human}.txt" =>
+			// "1629418506846614551-con1629418506846614551sul-2021-08-19_20-13-08-322108785.txt"
+			inputFilename: "${timestamp-unix}-con${timestamp-unix}sul-${timestamp-human}.txt",
+			outputFilenamePattern: `\d+-con\d+sul-\d{4}-\d{2}-\d{2}_\d{2}-\d{2}-\d{2}-\d{9}.txt`,
+		},
+		{
+			desc: "template valid but no known variables used",
+			// Test whether a Unix timestamp is injected if a filename template is used,
+			// but no known timestamp variables (timestamp-unix, timestamp-human) are
+			// referenced.
+			// For example:
+			// "{1+3}" => "4-1629417801659820807.log"
+			inputFilename: "${1+3}",
+			outputFilenamePattern: `4-\d+.log`,
+		},
+		{
+			desc: "template invalid unknown variable",
+			// Test whether an error occurs when an unknown variable is referenced
+			// in a template filename.
+			// For example:
+			// "${abc}-consul" => error
+			inputFilename: "${abc}-consul",
+			expectError: "failed to evaluate log filename template",
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.desc, func(t *testing.T) {
+			require := require.New(t)
+
+			tempDir := testutil.TempDir(t, t.Name())
+			logFile := LogFile{
+				fileName: c.inputFilename,
+				logPath:  tempDir,
+				MaxBytes: 10,
+				MaxFiles: -1,
+			}
+
+			n, err := logFile.Write([]byte("[INFO] Hello World"))
+			if c.expectError != "" {
+				require.Equal(n, 0)
+				require.Contains(err.Error(), c.expectError)
+			} else {
+				// Should be one file in the temporary directory
+				logFiles := listDir(t, tempDir)
+				require.Len(listDir(t, tempDir), 1)
+
+				// The filename should match the specified pattern
+				matched, err := regexp.Match(c.outputFilenamePattern, []byte(logFiles[0]))
```

Comment:
We already have a testing shim for `time.Now`. Instead of a regex here, I think it would be great to patch that shim, and that way all the tests can assert an exact value, instead of a regex. That should make the tests more strict, and also easier to read.

I wasn't able to find an example of that patching, but it should look roughly like this:

```go
now = func() time.Time {
    return time.Date(2021, 1, 2, 3, 4, 5, 0, time.UTC)
}
t.Cleanup(func() {
    now = time.Now
})
```

Reply:
This is what I was hoping for, but I thought my only option was dependency injection of the time module and passing in a mock (which I wasn't going to do). This looks much easier - I'll try it!